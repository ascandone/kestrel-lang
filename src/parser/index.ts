export { parse, unsafeParse, ParseResult } from "./parser";
