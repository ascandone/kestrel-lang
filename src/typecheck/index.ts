export {
  typecheck,
  typecheckProject,
  UnifyErrorType,
  TypecheckError,
  ProjectTypeCheckResult,
} from "./typecheck";
export { typeErrorPPrint, typePPrint } from "./pretty-printer";
