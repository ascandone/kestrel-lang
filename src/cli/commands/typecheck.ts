import { readFileSync } from "fs";
import { parse } from "../../parser";
import { typecheck } from "../../typecheck/typecheck";
import { typeErrorPPrint } from "../../typecheck/pretty-printer";
import { prelude } from "../../typecheck/prelude";
import { Span } from "../../ast";

const FgRed = "\x1b[31m";
const Reset = "\x1b[0m";
const FgBlack = "\x1b[30m";
const BgWhite = "\x1b[47m";

export function typecheckCmd(path: string) {
  const f = readFileSync(path);

  const src = f.toString();
  const parseResult = parse(src);
  if (!parseResult.ok) {
    console.log(
      `${FgRed}Parsing error:${Reset} ${parseResult.matchResult.message!}`,
    );
    return;
  }

  const [, errors] = typecheck(parseResult.value, prelude);

  for (const error of errors) {
    const msg = typeErrorPPrint(error);
    console.log(`${FgRed}Error:${Reset} ${msg}`);
    console.log(showErrorLine(src, error.node.span), "\n\n");
  }

  if (errors.length > 0) {
    const plErr = errors.length === 1 ? "error" : "errors";
    console.log(`[Found ${errors.length} ${plErr}]\n`);
  } else {
    console.log("Found no errors âœ…");
  }
}

export function showErrorLine(src: string, [start, end]: Span): string {
  const startPos = offsetToPosition(src, start);
  const endPos = offsetToPosition(src, end);
  const lines = src.split("\n");

  // TODO handle nums with many digits
  const digitsPadding = repeatN(" ", 1);

  function showLine(line: number) {
    const codeLine = lines[line]!;
    const lineNum = `${BgWhite}${FgBlack}${line + 1}${Reset}`;
    return `${lineNum} ${codeLine}`;
  }

  function showErr([start, end]: [start: number, end: number]) {
    const errPadding = repeatN(" ", start);
    const errHighlight = repeatN("~", end - start);
    return `${BgWhite}${digitsPadding}${Reset} ${FgRed}${errPadding}${errHighlight}${Reset}`;
  }

  const ret: string[] = [];
  for (
    let currentLine = startPos.line;
    currentLine <= endPos.line;
    currentLine++
  ) {
    const startChar = currentLine === startPos.line ? startPos.character : 0;

    const endChar =
      currentLine === endPos.line
        ? endPos.character
        : lines[currentLine]!.length;

    ret.push(showLine(currentLine), showErr([startChar, endChar]));
  }

  return ret.join("\n");
}

type Position = { line: number; character: number };

function offsetToPosition(src: string, offset: number): Position {
  let line = 0,
    character = 0;

  for (const ch of src) {
    if (offset === 0) {
      break;
    }

    offset--;
    if (ch === "\n") {
      line++;
      character = 0;
    } else {
      character++;
    }
  }

  return { line, character };
}

function repeatN(ch: string, times: number) {
  return Array.from({ length: times }, () => ch).join("");
}
