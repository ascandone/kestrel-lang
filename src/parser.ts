/* eslint-disable @typescript-eslint/no-unused-vars */
import grammar from "./parser/grammar.ohm-bundle";
import {
  ConstLiteral,
  Program,
  Expr,
  Declaration,
  Span,
  SpanMeta,
  TypeAst,
  TypeDeclaration,
  TypeVariant,
  MatchPattern,
} from "./ast";
import type {
  IterationNode,
  MatchResult,
  NonterminalNode,
  Node as OhmNode,
  TerminalNode,
} from "ohm-js";
import { TypeMeta } from "./typecheck/typecheck";

function getSpan({ source }: OhmNode): Span {
  return [source.startIdx, source.endIdx];
}

const semantics = grammar.createSemantics();

semantics.addOperation<ConstLiteral>("lit()", {
  number_whole(node) {
    return { type: "int", value: Number(node.sourceString) };
  },

  number_fract(_intPart, _comma, _floatPart) {
    return { type: "float", value: Number(this.sourceString) };
  },

  string(_lDel, _s, _rDel) {
    return { type: "string", value: JSON.parse(this.sourceString) };
  },
});

function parseInfix(
  this: NonterminalNode,
  left: NonterminalNode,
  op: TerminalNode,
  right: NonterminalNode,
): Expr<SpanMeta> {
  return {
    type: "application",
    span: getSpan(this),
    caller: {
      type: "identifier",
      name: op.sourceString,
      span: getSpan(op),
    },
    args: [left.expr(), right.expr()],
  };
}

function parsePrefix(
  this: NonterminalNode,
  op: TerminalNode,
  exp: NonterminalNode,
): Expr<SpanMeta> {
  return {
    type: "application",
    span: getSpan(this),
    caller: {
      type: "identifier",
      name: op.sourceString,
      span: getSpan(op),
    },
    args: [exp.expr()],
  };
}

semantics.addOperation<{ name: string } & SpanMeta>("ident()", {
  infixIdent(_lparens, ident, _rparens) {
    return {
      name: ident.sourceString,
      span: getSpan(this),
    };
  },
  ident(_hd, _tl) {
    return {
      name: this.sourceString,
      span: getSpan(this),
    };
  },
});

semantics.addOperation<MatchPattern>("matchPattern()", {
  MatchPattern_ident(ident) {
    return {
      type: "ident",
      ident: ident.sourceString,
      span: getSpan(this),
    };
  },
  MatchPattern_lit(lit) {
    return {
      type: "lit",
      literal: lit.lit(),
      span: getSpan(this),
    };
  },

  ConsPattern_cons(l, cons, r) {
    return {
      type: "constructor",
      name: cons.sourceString,
      args: [l.matchPattern(), r.matchPattern()],
      span: getSpan(this),
    };
  },

  MatchPattern_tuple(_lparens, x, _comma, xs, _rparens) {
    const xs_ = xs.asIteration().children.map((c) => c.matchPattern());

    const count = 1 + xs_.length;
    const name = `Tuple${count}`;
    return {
      type: "constructor",
      name,
      args: [x.matchPattern(), ...xs_],
      span: getSpan(this),
    };
  },
  MatchPattern_constructor(ident, _lparent, args, _rparens) {
    let args_: MatchPattern[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((a) => a.matchPattern());
    }
    return {
      type: "constructor",
      name: ident.sourceString,
      args: args_,
      span: getSpan(this),
    };
  },
});

semantics.addOperation<[MatchPattern, Expr<TypeMeta>]>("matchClause()", {
  MatchClause_clause(match, _arrow, expr) {
    return [match.matchPattern(), expr.expr()];
  },
});

semantics.addOperation<Expr<SpanMeta>>("expr()", {
  PipeExp_pipe(left, _pipe, caller, _lparens, args, _trailingComma, _rparens) {
    return {
      type: "application",
      caller: caller.expr(),
      args: [left.expr(), ...args.asIteration().children.map((c) => c.expr())],
      span: getSpan(this),
    };
  },
  PriExp_literal(n) {
    return {
      type: "constant",
      value: n.lit(),
      span: getSpan(this),
    };
  },

  ConsExpr_cons(hd, cons, tl) {
    return {
      type: "application",
      span: getSpan(this),
      caller: {
        type: "identifier",
        name: "Cons",
        span: getSpan(cons),
      },
      args: [hd.expr(), tl.expr()],
    };
  },

  PriExp_tuple(_lparens, x, _comma, xs, _rparens) {
    const xs_ = xs.asIteration().children.map((x) => x.expr());

    const count = 1 + xs_.length;

    return {
      type: "application",
      caller: {
        type: "identifier",
        name: `Tuple${count}`,
        span: getSpan(this),
      },
      args: [x.expr(), ...xs_],
      span: getSpan(this),
    };
  },
  PriExp_paren(_open, e, _close) {
    return e.expr();
  },
  PriExp_constructor(node) {
    return {
      type: "identifier",
      name: node.sourceString,
      span: getSpan(node),
    };
  },

  ident(_hd, _tl) {
    return {
      type: "identifier",
      ...this.ident(),
    };
  },
  CompExp_lt: parseInfix,
  CompExp_gt: parseInfix,
  CompExp_gte: parseInfix,
  CompExp_lte: parseInfix,
  EqExpr_eq: parseInfix,
  EqExpr_neq: parseInfix,
  OrExpr_or: parseInfix,
  AndExpr_and: parseInfix,
  AddExp_plus: parseInfix,
  AddExp_plusFloat: parseInfix,
  AddExp_minus: parseInfix,
  AddExp_minusFloat: parseInfix,
  MulExp_mult: parseInfix,
  MulExp_multFloat: parseInfix,
  MulExp_divide: parseInfix,
  MulExp_divideFloat: parseInfix,
  MulExp_rem: parseInfix,
  ExpExp_power: parseInfix,
  AddExp_concat: parseInfix,

  NotExp_not: parsePrefix,

  PriExp_apply(f, _lpar, args, _trailingComma, _rpar) {
    return {
      type: "application",
      caller: f.expr(),
      args: args.asIteration().children.map((c) => c.expr()),
      span: getSpan(this),
    };
  },

  PriExp_block(_lbracket, block, _rbracket) {
    return block.expr();
  },

  PriExp_fn(_fn, params, _lbracket, block, _rbracket) {
    return {
      type: "fn",
      params: params.asIteration().children.map((c) => c.ident()),
      body: block.expr(),
      span: getSpan(this),
    };
  },

  PriExp_if(_if, cond, _lb1, x, _rb1, _else, _lb2, y, _rb2) {
    return {
      type: "if",
      condition: cond.expr(),
      then: x.expr(),
      else: y.expr(),
      span: getSpan(this),
    };
  },

  PriExp_match(_match, expr, _lbracket, clauses, _comma, _rbracket) {
    return {
      type: "match",
      expr: expr.expr(),
      clauses: clauses.asIteration().children.map((c) => c.matchClause()),
      span: getSpan(this),
    };
  },

  BlockContent_exp(e) {
    return e.expr();
  },

  BlockContent_monadicLet(_let, bindFIdent, ident, _eq, value, _comma, body) {
    return {
      type: "application",
      caller: bindFIdent.expr(),
      args: [
        value.expr(),
        {
          type: "fn",
          params: [ident.ident()],
          body: body.expr(),
          span: getSpan(this),
        },
      ],
      span: getSpan(this),
    };
  },

  BlockContent_let(_let, ident, _eq, value, _comma, body) {
    return {
      type: "let",
      binding: ident.ident(),
      value: value.expr(),
      body: body.expr(),
      span: getSpan(this),
    };
  },
});

semantics.addOperation<TypeAst>("type()", {
  Type_any(_underscore) {
    return { type: "any", span: getSpan(this) };
  },

  Type_var(ident) {
    return { type: "var", ident: ident.sourceString, span: getSpan(this) };
  },

  Type_fn(_fn, _lparens, args, _rparens, _arrow, ret) {
    const args_ = args.asIteration().children.map((arg) => arg.type());
    return { type: "fn", args: args_, return: ret.type(), span: getSpan(this) };
  },

  Type_named(ident, _lbracket, args, _rbracket) {
    let args_: TypeAst[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((c) => c.type());
    }
    return {
      type: "named",
      args: args_,
      name: ident.sourceString,
      span: getSpan(this),
    };
  },
});

type Statement =
  | { type: "typeDeclaration"; decl: TypeDeclaration }
  | { type: "declaration"; decl: Declaration };

semantics.addOperation<TypeVariant>("typeVariant()", {
  TypeVariant(name, _lparens, args, _rparens) {
    let args_: TypeAst[] = [];
    if (args.numChildren > 0) {
      args_ = args
        .child(0)
        .asIteration()
        .children.map((c) => c.type());
    }

    return { name: name.sourceString, args: args_ };
  },
});

function parseParams(params: IterationNode) {
  let params_: Array<{ name: string } & SpanMeta> = [];
  if (params.numChildren > 0) {
    params_ = params
      .child(0)
      .asIteration()
      .children.map((c) => ({
        name: c.sourceString,
        span: getSpan(c),
      }));
  }
  return params_;
}
semantics.addOperation<Statement>("statement()", {
  TypeDeclaration_externType(_extern, _type, typeName, _lT, params, _rT) {
    return {
      type: "typeDeclaration",
      decl: {
        type: "extern",
        params: parseParams(params),
        name: typeName.sourceString,
        span: getSpan(this),
      },
    };
  },

  TypeDeclaration_typeDef(
    _type,
    typeName,
    _lT,
    params,
    _rT,
    _lbracket,
    variants,
    _trailingComma,
    _rbracket,
  ) {
    const variants_ = variants
      .asIteration()
      .children.map<TypeVariant>((n) => n.typeVariant());

    return {
      type: "typeDeclaration",
      decl: {
        type: "adt",
        params: parseParams(params),
        name: typeName.sourceString,
        variants: variants_,
        span: getSpan(this),
      },
    };
  },
  Declaration_externLetStmt(_extern, _let, ident, _colon, typeHint) {
    return {
      type: "declaration",
      decl: {
        extern: true,
        binding: ident.ident(),
        span: getSpan(this),
        typeHint: {
          ...typeHint.type(),
          span: getSpan(typeHint.child(0)),
        },
      },
    };
  },
  Declaration_letStmt(_let, ident, _colon, typeHint, _eq, exp) {
    const th =
      typeHint.numChildren === 0
        ? {}
        : {
            typeHint: {
              ...typeHint.child(0).type(),
              span: getSpan(typeHint.child(0)),
            },
          };

    return {
      type: "declaration",
      decl: {
        extern: false,
        binding: ident.ident(),
        value: exp.expr(),
        span: getSpan(this),
        ...th,
      },
    };
  },
});

semantics.addOperation<Program>("parse()", {
  MAIN(statements) {
    const statements_ = statements.children.map<Statement>((child) =>
      child.statement(),
    );

    return {
      typeDeclarations: statements_.flatMap((st) =>
        st.type === "typeDeclaration" ? [st.decl] : [],
      ),
      declarations: statements_.flatMap((st) =>
        st.type === "declaration" ? [st.decl] : [],
      ),
    };
  },
});

export type ParseResult<T> =
  | { ok: true; value: T }
  | { ok: false; matchResult: MatchResult };

export function parse(input: string): ParseResult<Program<SpanMeta>> {
  const matchResult = grammar.match(input);
  if (matchResult.failed()) {
    return { ok: false, matchResult };
  }

  return { ok: true, value: semantics(matchResult).parse() };
}

export function unsafeParse(input: string): Program<SpanMeta> {
  const res = parse(input);
  if (res.ok) {
    return res.value;
  }

  throw new Error(res.matchResult.message!);
}
