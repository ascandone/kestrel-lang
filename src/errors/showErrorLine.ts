import { Span } from "../parser";
import { col } from "../utils/colors";

type Position = { line: number; character: number };

function repeatN(ch: string, times: number) {
  return Array.from({ length: times }, () => ch).join("");
}

export function showErrorLine(src: string, [start, end]: Span): string {
  const startPos = offsetToPosition(src, start);
  const endPos = offsetToPosition(src, end);
  const lines = src.split("\n");

  function showLine(line: number) {
    const codeLine = lines[line]!;
    return `${line + 1}| ${codeLine}`;
  }

  function showErr(
    currentLine: number,
    [start, end]: [start: number, end: number],
  ) {
    const lineDigits = currentLine.toString().length;
    const digitsPadding = repeatN(" ", lineDigits);

    const errPadding = repeatN(" ", start);
    const errHighlight = repeatN("~", end - start);
    return `${digitsPadding}| ${col.red.str(errPadding + errHighlight)}`;
  }

  const ret: string[] = [];
  for (
    let currentLine = startPos.line;
    currentLine <= endPos.line;
    currentLine++
  ) {
    const startChar = currentLine === startPos.line ? startPos.character : 0;

    const endChar =
      currentLine === endPos.line
        ? endPos.character
        : lines[currentLine]!.length;

    ret.push(showLine(currentLine), showErr(currentLine, [startChar, endChar]));
  }

  return ret.join("\n");
}

function offsetToPosition(src: string, offset: number): Position {
  let line = 0,
    character = 0;

  for (const ch of src) {
    if (offset === 0) {
      break;
    }

    offset--;
    if (ch === "\n") {
      line++;
      character = 0;
    } else {
      character++;
    }
  }

  return { line, character };
}
