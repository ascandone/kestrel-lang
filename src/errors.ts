import { showErrorLine } from "./errors/showErrorLine";
import { Span } from "./parser";
import { Type, typeToString } from "./typecheck";
import { col, withDisabled } from "./utils/colors";

export type Severity = "error" | "warning";

export type ErrorInfo = {
  span: Span;
  description: ErrorDescription;
};

export interface ErrorDescription {
  severity: Severity;
  errorName: string;
  shortDescription(): string;
}

export class ParsingError implements ErrorDescription {
  constructor(public expecting: string) {}
  severity: Severity = "error";
  errorName = "Parsing error";
  shortDescription(): string {
    return this.expecting;
  }
}

export class InvalidPipe implements ErrorDescription {
  constructor() {}
  severity: Severity = "error";
  errorName = "Invalid pipe";
  shortDescription(): string {
    return "Pipe right side should be a function application";
  }
}

export class DuplicateDeclaration implements ErrorDescription {
  constructor(public ident: string) {}
  severity: Severity = "error";
  errorName = "Duplicate declaration";
  shortDescription() {
    return `"${this.ident}" was already defined`;
  }
}

export class UnboundVariable implements ErrorDescription {
  constructor(public ident: string) {}
  severity: Severity = "error";
  errorName = "Unbound variable";
  shortDescription() {
    return `Cannot find variable "${this.ident}"`;
  }
}

export class CyclicDefinition implements ErrorDescription {
  constructor(public ident: string) {}
  severity: Severity = "error";
  errorName = "Cyclic definition";
  // TODO better error
  shortDescription() {
    return `Cyclic definition`;
  }
}

export class UnusedVariable implements ErrorDescription {
  constructor(
    public ident: string,
    public type: "local" | "global",
  ) {}

  severity: Severity = "warning";

  errorName = "Unused variable";
  shortDescription() {
    const pre = `"${this.ident}" is declared but never used.`;
    switch (this.type) {
      case "local":
        return `${pre}\nTry to rename it to "_${this.ident}"`;
      case "global":
        return `${pre}\nTry to add a \`pub\` modifier`;
    }
  }
}

export class UnusedImport implements ErrorDescription {
  severity: Severity = "warning";
  errorName: string = "Unused import";
  constructor(public namespace: string) {}
  shortDescription(): string {
    return `Import is never used: ${this.namespace}`;
  }
}

export class UnusedExposing implements ErrorDescription {
  severity: Severity = "warning";
  errorName: string = "Unused import";
  constructor(public name: string) {}
  shortDescription(): string {
    return `Exposed value is never used: "${this.name}"`;
  }
}

export class UnboundType implements ErrorDescription {
  constructor(public ident: string) {}
  severity: Severity = "error";
  errorName = "Unbound type";
  shortDescription() {
    return `Cannot find type "${this.ident}"`;
  }
}

export class InvalidTypeArity implements ErrorDescription {
  constructor(
    public type: string,
    public expected: number,
    public got: number,
  ) {}
  severity: Severity = "error";
  errorName = "Invalid type arity";
  shortDescription() {
    return `Wrong number of args for type "${this.type}". Expected ${this.expected} but got ${this.got} instead`;
  }
}

export class UnboundTypeParam implements ErrorDescription {
  constructor(public param: string) {}
  severity: Severity = "error";
  errorName = "Unbound type parameter";
  shortDescription() {
    return `Cannot find type parameter "${this.param}"`;
  }
}

export class InvalidCatchall implements ErrorDescription {
  errorName = "Invalid catchall";
  severity: Severity = "error";
  shortDescription() {
    return `Invalid use of the catchall type`;
  }
}

export class TypeParamShadowing implements ErrorDescription {
  constructor(public param: string) {}
  severity: Severity = "error";
  errorName = "Type parameter shadowing";
  shortDescription(): string {
    return `Cannot redeclare type parameter ${this.param}`;
  }
}

export class ArityMismatch implements ErrorDescription {
  constructor(
    public expected: number,
    public got: number,
  ) {}
  severity: Severity = "error";
  errorName = "Arity mismatch";
  shortDescription(): string {
    return `Expected ${this.expected} arguments, but got ${this.got}.`;
  }
}

export class UnboundModule implements ErrorDescription {
  constructor(public moduleName: string) {}
  severity: Severity = "error";
  errorName = "Unbound module";
  shortDescription(): string {
    return `Unbound module: "${this.moduleName}".`;
  }
}

export class UnimportedModule implements ErrorDescription {
  constructor(public moduleName: string) {}
  severity: Severity = "error";
  errorName = "Unimported module";
  shortDescription(): string {
    return `This module was not imported: "${this.moduleName}".`;
  }
}

export class NonExistingImport implements ErrorDescription {
  constructor(public name: string) {}
  severity: Severity = "error";
  errorName = "Non existing import";
  shortDescription(): string {
    return `The module does not expose the following value: ${this.name}`;
  }
}

export class BadImport implements ErrorDescription {
  errorName = "Bad import";
  severity: Severity = "error";
  shortDescription(): string {
    return `This type doesn't have constructors to expose`;
  }
}

export class OccursCheck implements ErrorDescription {
  errorName = "Occurs check";
  severity: Severity = "error";
  shortDescription(): string {
    return `Cannot construct the infinite type`;
  }
}

export class NonExhaustiveMatch implements ErrorDescription {
  severity: Severity = "error";
  errorName: string = "Non exhaustive match";
  shortDescription(): string {
    return "Match is not exhaustive";
  }
}

export class TraitNotSatified implements ErrorDescription {
  severity: Severity = "error";

  errorName: string = "Trait not satisfied";

  constructor(
    public readonly type: Type,
    public readonly trait: string,
  ) {}

  shortDescription(): string {
    const type = typeToString(this.type);
    return `Cannot satisfy trait ${this.trait} for type ${type}`;
  }
}

export class TypeMismatch implements ErrorDescription {
  constructor(
    public expected: Type,
    public got: Type,
  ) {}

  errorName = "Type mismatch";
  severity: Severity = "error";
  shortDescription(): string {
    const expected = typeToString(this.expected);
    const got = typeToString(this.got);
    const qualify = expected === got;

    const nsLeft =
      qualify && this.expected.type === "named"
        ? `${this.expected.moduleName}.`
        : "";

    const nsRight =
      qualify && this.got.type === "named" ? `${this.got.moduleName}.` : "";

    return `Expected:  ${nsLeft}${expected}
     Got:  ${nsRight}${got}
`;
  }
}

export class InvalidField implements ErrorDescription {
  severity: Severity = "error";
  errorName: string = "Invalid field";

  constructor(
    public type: string,
    public field: string,
  ) {}

  shortDescription(): string {
    return `The field '${this.field}' does not exist on type '${this.type}'`;
  }
}

export class MissingRequiredFields implements ErrorDescription {
  severity: Severity = "error";
  errorName: string = "Missing required field";

  constructor(
    public type: string,
    public fields: string[],
  ) {}

  shortDescription(): string {
    if (this.fields.length === 1) {
      return `Missing field '${this.fields[0]}' (required on type '${this.type}')`;
    }

    const missingFields = this.fields.map((f) => `'${f}'`).join(", ");
    return `Missing the following fields: ${missingFields} (required on type '${this.type}')`;
  }
}

export function errorInfoToString(
  src: string,
  { description, span }: ErrorInfo,
  disableColors: boolean = false,
): string {
  return withDisabled(disableColors, () => {
    const sev = description.severity ?? "error";
    const errType =
      sev === "error" ? col.red.tag`Error:` : col.yellow.tag`Warning:`;

    return `${errType} ${col.bright.str(description.errorName)}

${description.shortDescription()}

${showErrorLine(src, span)}`;
  });
}
