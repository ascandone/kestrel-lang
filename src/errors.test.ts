import { describe, expect, test } from "vitest";
import { errorInfoToString } from "./errors";
import { unsafeParse } from "./parser";
import { typecheck } from "./typecheck";

describe(errorInfoToString.name, () => {
  test("unbound variable", () => {
    const src = "pub let x = unbound_var";
    snapshotErr(src);
  });

  test("arity mismatch (too many args)", () => {
    const src = `
      pub let f = |x| { x }
      pub let x = f(0, 1)
    `;
    snapshotErr(src);
  });

  test("arity mismatch (too few args)", () => {
    const src = `
      pub let f = |x, _y| { x }
      pub let x = f(0)
    `;
    snapshotErr(src);
  });

  test("type error", () => {
    const src = `
      type Custom {}
      pub let x: Custom = "abc"
    `;
    snapshotErr(src);
  });
});

function snapshotErr(src: string, opts: { DEBUG_LOG?: boolean } = {}) {
  const module = "Main";
  const parsed = unsafeParse(src);
  const [, errors] = typecheck(module, parsed, {}, []);

  if (errors.length !== 1) {
    console.log(errors);
  }
  expect(errors).toHaveLength(1);

  const error = errors[0]!;
  const out = errorInfoToString(src, error);

  if (opts.DEBUG_LOG) {
    console.log(out);
  }

  expect(out).toMatchSnapshot();
}
