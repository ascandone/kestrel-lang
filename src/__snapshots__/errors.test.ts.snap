// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`errorInfoToString > unbound variable 1`] = `
"[31mError:[0m [1mUnbound variable[0m

Cannot find variable "unbound_var"

1| pub let x = unbound_var
 | [31m            ~~~~~~~~~~~[0m"
`;
